### DISCUSSION TITLE
How do you handle multiple repositories?

### DISCUSSION THREAD

**Discussion by uris77** [benbjohnson/wtf#47] [Q&A] [answered]
I often find myself working in applications where we have a `search database` (elastic, algolia, etc) and the `source of truth` database (Postgres, firebase, etc). The general flow is that when a record is created, that record is also duplicated in the search database. This is done to facilitate full text search (among other things like phonetic search, etc).

How would you organize something like this? Although sometimes the services would look the same, they differ slightly:

```
// PersonService represents a service for managing persons 
type PersonService interface {
	// CreatePerson creates a new person.
	CreatePerson(ctx context.Context, person models.Person) error

	// UpdatePerson updates the person's arrival and personal information.
	// Note that screening and vaccination information is updated separately.
	UpdatePerson(ctx context.Context, person models.Person) error

       // SearchPerson conducts a full text search for a person. The txt could match the full name,
      // or any part of the name, as well as their QR code, etc.
     SearchPerson(ctx context.Context, txt string) ([]models.Person, error)
     
     // GetByID retrieves a person record with the matching id
    GetByID(cox context.Context, id string) (models.Person, error)
}
```

In this example, the `Postgres` database would only implement `CreatePerson`, `UpdatePerson`, & `GetByID`, while
the search database would implement `CreatePerson`, `UpdatePerson` & `SearchPerson`. 

1. Do I need different interfaces for each type of database?
2. Is it a good idea to use only this interface `PersonService` and only implement the methods that relates to each type of database?

**Comment by benbjohnson** [comment, marked as answer]
@uris77 Good question. I haven't found a perfect way to separate the database layer & the search layer personally so take what I say with a grain of salt. If you can fit your search service into the same interface then I would do that. Sometimes it's tough since your search service may only return a subset of fields (or just the ID) but in that case you can just return a `Person` with those subset of fields.

From there, I will usually attach the search service to my main database service:

```go
package postgres

type PersonService struct {
	db *sql.DB

	SearchService PersonService // set this to elastic.PersonService in main
}
```

Then your Postgres implementation can delegate calls to the search service as needed. If the search service doesn't fit well into the same interface, though, then I wouldn't try to force it.

If you have some methods that are not implemented, then I'll typically return an application-specific `ErrNotImplemented` to indicate that it shouldn't be used.
